Enhance progress bar

aaronlippold opened the pull request:
- accounted for profile errors at runtime and report them as [ERROR]s
- accounted for `impact 0.0` controls report them as [NA] - aka Not Applicable
- added the control title to the output for end-user context
- added the control status to help those who can't see color easily
- fixed control status to account for passed,skipped and passed,skipped,failed

aaronlippold committed:
enhanced the progress reporter

- accounted for profile errors at runtime
- accounted for `impact 0.0` controls and set them to NA
- added the control title to the output for end-user context

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
Revert "Bump version to 5.7.8 by Chef Expeditor"

This reverts commit be0b8fd10904b68ff5ebbd3af981e963be36b1ed.

aaronlippold committed:
Revert "enhanced the progress reporter"

This reverts commit 5c71df6cf16043b2ffad09509a72ad9df752f9f8.

aaronlippold committed:
Enhanced Progress Bar Reporter

- accounted for profile errors at run-time - added example_error
- accounted for NA controls with impact 0.0 - added example_na
- added control title to output for end-user clarity
- added control status text to output for color bindness / user ease

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
linting with rake test:lint:auto_correct

Signed-off-by: Aaron Lippold <lippold@gmail.com>

Nik08 commented on lib/plugins/inspec-streaming-reporter-progress-bar/lib/inspec-streaming-reporter-progress-bar/streaming_reporter.rb:
```diff
@@ -9,104 +10,144 @@ module InspecPlugins::StreamingReporterProgressBar
       indicator = INDICATORS[control_status]
-      message_to_format = ""
-      message_to_format += "#{indicator}  "
-      message_to_format += control_id.to_s.lstrip.force_encoding(Encoding::UTF_8)
```
`force_encoding(Encoding::UTF_8)` we should probably keep this, this is a safe check.

aaronlippold committed:
- fixed frozen_string exception - gsub! was the issue.
- put encoding back into place
- added exception logging
- tested with two and three level dependant profiles
- fixed all unit tests

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
linting

Signed-off-by: Aaron Lippold <lippold@gmail.com>

Nik08 commented on lib/plugins/inspec-streaming-reporter-progress-bar/lib/inspec-streaming-reporter-progress-bar/streaming_reporter.rb:
```diff
@@ -28,85 +32,128 @@ class StreamingReporter < Inspec.plugin(2, :streaming_reporter)
+      message_to_format += "#{full_description.gsub(/\n*\s+/, " ").to_s.force_encoding(Encoding::UTF_8)}  " unless title
       format_with_color(control_status, message_to_format)
+    rescue StandardError => e
+      raise "Exception in show_progress: #{e}"
```
I think we should have error messages that could be also understood by an inspec user. 
So putting `show_progress` and `format_with_color` does help a developer to debug. But maybe not so useful for the user. Maybe use a message that is useful to both user and a developer? Makes sense?

aaronlippold replied:
Happy to, what kind of message should we provide do you think? Parse `e` to bubble up the issue or put a case statment in that captures the N common errors that can happen at this stage and 'decode' them into nature language for the end user.

aaronlippold replied:
Do you have a few examples of other places in the code base where you have done this so I could use them as a reference?

ejaronne replied:
> I think we should have error messages that could be also understood by an inspec user.

When it comes to InSpec profiles, we use the following short phrases to qualify what each outcome means to the user, so they are not confused as to what has happened and "whose fault is it?":

Passed: "All sub-tests for this control have passed or were skipped"
Failed: "At least one sub-test failed for this control"
Not Applicable: "System exception or absent component"
Not Reviewed: "Can only be tested manually at this time"
Profile Error: "Errors running test - check profile run privileges or check with the author of profile."

Could these somehow be added to the enhanced reporting engine?  Thanks!

Nik08 replied:
@aaronlippold Before you made this change what I had here was `Exception in Progress Bar streaming reporter`. Nothing fancy but somehow clear to the user where it went wrong? So maybe mix of this and the new change of message that you have made. Sounds good? 

And for examples, you can search with `exception` keyword in this repo and you will get an idea how we have framed messages in case of exceptions.

aaronlippold committed:
commits

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
Merge remote-tracking branch 'origin/main' into enhance-progress-bar

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
added the advanced reporting flag and put original status back by default

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
added the advanced reporting flag and put original status back by default

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
fixing conflict and merge from main

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
Merge remote-tracking branch 'upstream/main' into enhance-progress-bar

Signed-off-by: Aaron Lippold <lippold@gmail.com>
